### ISSUE TITLE
Installation Failure Due to `py_find_1st` Package on Ubuntu 24.04

### ISSUE THREAD

**Issue by ASGDLO** [freqtrade/freqtrade#10350]
---

### Issue: Installation Failure Due to `py_find_1st` Package on Ubuntu 24.04

**Environment:**
- **OS:** Ubuntu 24.04
- **Python Version:** 3.12
- **Freqtrade Version:** 2024.5

**Description:**
Encountering an issue while installing `freqtrade` on Ubuntu 24.04. The installation fails due to a problem with the `py_find_1st` package. The error occurs during the extraction of the tarball for `py_find_1st` version 1.1.6.

**Error Message:**
```
ERROR: Exception:
Traceback (most recent call last):
  File "/home/ethan/Documents/freqtrade/.venv/lib/python3.12/site-packages/pip/_internal/cli/base_command.py", line 179, in exc_logging_wrapper
    status = run_func(*args)
  File "/home/ethan/Documents/freqtrade/.venv/lib/python3.12/site-packages/pip/_internal/cli/req_command.py", line 67, in wrapper
    return func(self, options, args)
  File "/home/ethan/Documents/freqtrade/.venv/lib/python3.12/site-packages/pip/_internal/commands/install.py", line 377, in run
    requirement_set = resolver.resolve(
  File "/home/ethan/Documents/freqtrade/.venv/lib/python3.12/site-packages/pip/_internal/resolution/resolvelib/resolver.py", line 95, in resolve
    result = self._result = resolver.resolve(
  File "/home/ethan/Documents/freqtrade/.venv/lib/python3.12/site-packages/pip/_vendor/resolvelib/resolvers.py", line 546, in resolve
    state = resolution.resolve(requirements, max_rounds=max_rounds)
  File "/home/ethan/Documents/freqtrade/.venv/lib/python3.12/site-packages/pip/_vendor/resolvelib/resolvers.py", line 397, in resolve
    self._add_to_criteria(self.state.criteria, r, parent=None)
  File "/home/ethan/Documents/freqtrade/.venv/lib/python3.12/site-packages/pip/_vendor/resolvelib/resolvers.py", line 173, in _add_to_criteria
    if not criterion.candidates:
  File "/home/ethan/Documents/freqtrade/.venv/lib/python3.12/site-packages/pip/_vendor/resolvelib/structs.py", line 156, in __bool__
    return bool(self._sequence)
  File "/home/ethan/Documents/freqtrade/.venv/lib/python3.12/site-packages/pip/_internal/resolution/resolvelib/found_candidates.py", line 174, in __bool__
    return any(self)
  File "/home/ethan/Documents/freqtrade/.venv/lib/python3.12/site-packages/pip/_internal/resolution/resolvelib/found_candidates.py", line 162, in <genexpr>
    return (c for c in iterator if id(c) not in self._incompatible_ids)
  File "/home/ethan/Documents/freqtrade/.venv/lib/python3.12/site-packages/pip/_internal/resolution/resolvelib/found_candidates.py", line 53, in _iter_built
    candidate = func()
  File "/home/ethan/Documents/freqtrade/.venv/lib/python3.12/site-packages/pip/_internal/resolution/resolvelib/factory.py", line 185, in _make_candidate_from_link
    base: Optional[BaseCandidate] = self._make_base_candidate_from_link(
  File "/home/ethan/Documents/freqtrade/.venv/lib/python3.12/site-packages/pip/_internal/resolution/resolvelib/factory.py", line 231, in _make_base_candidate_from_link
    self._link_candidate_cache[link] = LinkCandidate(
  File "/home/ethan/Documents/freqtrade/.venv/lib/python3.12/site-packages/pip/_internal/resolution/resolvelib/candidates.py", line 303, in __init__
    super().__init__(
  File "/home/ethan/Documents/freqtrade/.venv/lib/python3.12/site-packages/pip/_internal/resolution/resolvelib/candidates.py", line 158, in __init__
    self.dist = self._prepare()
  File "/home/ethan/Documents/freqtrade/.venv/lib/python3.12/site-packages/pip/_internal/resolution/resolvelib/candidates.py", line 235, in _prepare
    dist = self._prepare_distribution()
  File "/home/ethan/Documents/freqtrade/.venv/lib/python3.12/site-packages/pip/_internal/resolution/resolvelib/candidates.py", line 314, in _prepare_distribution
    return preparer.prepare_linked_requirement(self._ireq, parallel_builds=True)
  File "/home/ethan/Documents/freqtrade/.venv/lib/python3.12/site-packages/pip/_internal/operations/prepare.py", line 527, in prepare_linked_requirement
    return self._prepare_linked_requirement(req, parallel_builds)
  File "/home/ethan/Documents/freqtrade/.venv/lib/python3.12/site-packages/pip/_internal/operations/prepare.py", line 598, in _prepare_linked_requirement
    local_file = unpack_url(
  File "/home/ethan/Documents/freqtrade/.venv/lib/python3.12/site-packages/pip/_internal/operations/prepare.py", line 180, in unpack_url
    unpack_file(file.path, location, file.content_type)
  File "/home/ethan/Documents/freqtrade/.venv/lib/python3.12/site-packages/pip/_internal/utils/unpacking.py", line 316, in unpack_file
    untar_file(filename, location)
  File "/home/ethan/Documents/freqtrade/.venv/lib/python3.12/site-packages/pip/_internal/utils/unpacking.py", line 235, in untar_file
    tar.extractall(location, filter=pip_filter)
  File "/usr/lib/python3.12/tarfile.py", line 2269, in extractall
    self._extract_one(tarinfo, path, set_attrs=not tarinfo.isdir(),
  File "/usr/lib/python3.12/tarfile.py", line 2332, in _extract_one
    self._extract_member(tarinfo, os.path.join(path, tarinfo.name),
  File "/usr/lib/python3.12/tarfile.py", line 2423, in _extract_member
    self.makelink(tarinfo, targetpath)
  File "/usr/lib/python3.12/tarfile.py", line 2521, in makelink
    self._extract_member(self._find_link_target(tarinfo),
  File "/usr/lib/python3.12/tarfile.py", line 2725, in _find_link_target
    raise KeyError("linkname 'py_find_1st-1.1.6/README' not found")
KeyError: "linkname 'py_find_1st-1.1.6/README' not found"
```

**Steps to Reproduce:**
1. Clone the `freqtrade` repository.
2. cd freqtrade
3. ./setup.sh -i

**Expected Behavior:**
The installation should complete successfully without errors.

**Actual Behavior:**
The installation fails with the above traceback, specifically due to an issue with the `py_find_1st` package.

**Temporary Workaround:**
Activating the virtual environment and manually installing `py_find_1st` with a different version fixes the issue:
```bash
source ./.venv/bin/activate
pip install py_find_1st==1.1.5
deactivate
./setup.sh -u
```

**Additional Information:**
This issue might be specific to the tarball extraction process on Ubuntu 24.04. Any assistance in resolving this problem would be appreciated.

**Suggested Fix:**
Consider updating the `py_find_1st` package or provide an alternative package/version that works correctly on Ubuntu 24.04.

Thank you!

**Comment by xmatthias** [comment]
Well it's a problem with `py_find_1st` - first and foremost.

We're using the latest version of that package - so there's no point in "updating" when there's no update.

While i understand that you've experienced this problem with freqtrade - you should report it at the source - the  [py_find_1st repository](https://github.com/roebel/py_find_1st/).
In my experience, the author is pretty responsive when it comes to this.

For now, i don't intend to take further action - please report this in the correct place, thanks.


Until he's got time to fix this (and release the fix) - i'd recommend to use docker - which will circumvent this and will be host-system independently working.

**Comment by xmatthias** [comment]
In a funny turn of events, this appears to be a bug (or regression or intentional, undocumented change) in pip's latest version.
https://github.com/pypa/pip/issues/12781

I'll hence keep this open for now - but for the moment, i don't see the next steps with freqtrade.

**Comment by xmatthias** [comment]
We've added a workaround for this for now - pinning pip to <=24.0 (24.1 breaks this).

that workaround will probably be removed once the upstream bug / problem has been fixed - though i'll be closing this here - as it's effectively no longer a problem for freqtrade at this point.

**Comment by kivwil** [comment]
Can you explain how to use the workaround?

I cannot make it work by using:
```
source ./.venv/bin/activate
pip install py_find_1st==1.1.5
```

**Comment by xmatthias** [comment]
You need to downgrade pip. 

`pip install pip==24.0` 

Its got nothing to do with py_find_1st.

**Comment by kivwil** [comment]
> You need to downgrade pip.
> 
> `pip install pip==24.0`
> 
> Its got nothing to do with py_find_1st.

I have pip 24.0 and Python 3.12.3.

In this case running `./setup.sh -i` fails with the same errors.

When I edit this line in setup.sh:
`${PYTHON} -m pip install --upgrade pip wheel setuptools`
to
`${PYTHON} -m pip install --upgrade pip==24.0 wheel setuptools`
it also fails:

```
Using cached pycparser-2.22-py3-none-any.whl (117 kB)
Using cached shellingham-1.5.4-py2.py3-none-any.whl (9.8 kB)
Building wheels for collected packages: TA-Lib
  Building wheel for TA-Lib (pyproject.toml) ... error
  error: subprocess-exited-with-error

  × Building wheel for TA-Lib (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [44 lines of output]
      running bdist_wheel
      running build
      running build_py
      creating build
      creating build/lib.linux-x86_64-cpython-312
      creating build/lib.linux-x86_64-cpython-312/talib
      copying talib/stream.py -> build/lib.linux-x86_64-cpython-312/talib
      copying talib/__init__.py -> build/lib.linux-x86_64-cpython-312/talib
      copying talib/abstract.py -> build/lib.linux-x86_64-cpython-312/talib
      copying talib/deprecated.py -> build/lib.linux-x86_64-cpython-312/talib
      running build_ext
      building 'talib._ta_lib' extension
      creating build/temp.linux-x86_64-cpython-312
      creating build/temp.linux-x86_64-cpython-312/talib
      x86_64-linux-gnu-gcc -fno-strict-overflow -Wsign-compare -DNDEBUG -g -O2 -Wall -fPIC -I/usr/include -I/usr/local/include -I/opt/include -I/opt/local/include -I/opt/homebrew/include -I/opt/homebrew/opt/ta-lib/include -I/tmp/pip-build-env-u6qcz2_5/normal/lib/python3.12/site-packages/numpy/_core/include -I/home/hk/apps/freqtrade/.venv/include -I/usr/include/python3.12 -c talib/_ta_lib.c -o build/temp.linux-x86_64-cpython-312/talib/_ta_lib.o
      In file included from /tmp/pip-build-env-u6qcz2_5/normal/lib/python3.12/site-packages/numpy/_core/include/numpy/ndarraytypes.h:1909,
                       from /tmp/pip-build-env-u6qcz2_5/normal/lib/python3.12/site-packages/numpy/_core/include/numpy/ndarrayobject.h:12,
                       from /tmp/pip-build-env-u6qcz2_5/normal/lib/python3.12/site-packages/numpy/_core/include/numpy/arrayobject.h:5,
                       from talib/_ta_lib.c:1231:
      /tmp/pip-build-env-u6qcz2_5/normal/lib/python3.12/site-packages/numpy/_core/include/numpy/npy_1_7_deprecated_api.h:17:2: warning: #warning "Using deprecated NumPy API, disable it with " "#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
         17 | #warning "Using deprecated NumPy API, disable it with " \
            |  ^~~~~~~
      In file included from /usr/include/python3.12/Python.h:38,
                       from talib/_ta_lib.c:16:
      talib/_ta_lib.c: In function ‘__pyx_f_5numpy_PyDataType_SHAPE’:
      talib/_ta_lib.c:10264:39: error: ‘PyArray_Descr’ {aka ‘struct _PyArray_Descr’} has no member named ‘subarray’
      10264 |     __Pyx_INCREF(((PyObject*)__pyx_v_d->subarray->shape));
            |                                       ^~
      /usr/include/python3.12/pyport.h:24:38: note: in definition of macro ‘_Py_CAST’
         24 | #define _Py_CAST(type, expr) ((type)(expr))
            |                                      ^~~~
      /usr/include/python3.12/object.h:661:35: note: in expansion of macro ‘_PyObject_CAST’
        661 | #  define Py_INCREF(op) Py_INCREF(_PyObject_CAST(op))
            |                                   ^~~~~~~~~~~~~~
      talib/_ta_lib.c:1653:27: note: in expansion of macro ‘Py_INCREF’
       1653 |   #define __Pyx_INCREF(r) Py_INCREF(r)
            |                           ^~~~~~~~~
      talib/_ta_lib.c:10264:5: note: in expansion of macro ‘__Pyx_INCREF’
      10264 |     __Pyx_INCREF(((PyObject*)__pyx_v_d->subarray->shape));
            |     ^~~~~~~~~~~~
      talib/_ta_lib.c:10265:36: error: ‘PyArray_Descr’ {aka ‘struct _PyArray_Descr’} has no member named ‘subarray’
      10265 |     __pyx_r = ((PyObject*)__pyx_v_d->subarray->shape);
            |                                    ^~
      error: command '/usr/bin/x86_64-linux-gnu-gcc' failed with exit code 1
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for TA-Lib
Failed to build TA-Lib
ERROR: Could not build wheels for TA-Lib, which is required to install pyproject.toml-based projects
Failed installing dependencies
```

Any solutions?

**Comment by stash86** [comment]
Use develop branch

**Comment by Rainking80** [comment]
I see the same error. Using develop branch and pip 24.0. I also switched to python 3.10. No success yet. Help would be appreciated.

Update:

Switching manually to py_find_1st==1.1.5 in reqirements.txt fixed the issue for me.

by the way changing python to 3.10 was a bad idea. i t crippled my os ;)

**Comment by stash86** [comment]
I have updated FT on 3 different vps with no issue (using develop branch).

Changing python surely isn't the solution nor the issue in the first place

**Comment by froggleston** [comment]
`it crippled my os` - this will never happen with a venv. It sounds like you were trying to apply the solution to your system python, not your venv, which is likely why it didn't work.

**Comment by froggleston** [comment]
As people are getting confused still, particularly with fresh installs using `setup.sh -i`, the easiest fix is to add `pip==24.0` to the `requirements.txt` and rerun the setup script.

**Comment by froggleston** [comment]
Pretty sure this is actually still a valid issue: https://github.com/freqtrade/freqtrade/pull/10374

**Comment by xmatthias** [comment]
Fixed in #10374
